pkg/monitors/jenkins/jenkins.go

```diff
@@ -0,0 +1,121 @@
+package jenkins
+
+import (
+	"context"
+	"time"
+
+	"github.com/signalfx/signalfx-agent/pkg/core/common/httpclient"
+	"github.com/signalfx/signalfx-agent/pkg/core/config"
+	"github.com/signalfx/signalfx-agent/pkg/monitors"
+	jc "github.com/signalfx/signalfx-agent/pkg/monitors/jenkins/client"
+	"github.com/signalfx/signalfx-agent/pkg/monitors/types"
+	"github.com/signalfx/signalfx-agent/pkg/utils"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	healthEndpoint  = "metrics/%s/healthcheck/"
+	metricsEndpoint = "metrics/%s/metrics/"
+	pingEndPoint    = "metrics/%s/ping/"
+)
+
+// Config for this monitor
+type Config struct {
+	config.MonitorConfig  `yaml:",inline" acceptsEndpoints:"true"`
+	httpclient.HTTPConfig `yaml:",inline"`
+
+	// Host of the jenkins instance
+	Host string `yaml:"host" validate:"required"`
+	// Port of the jenkins instance
+	Port string `yaml:"port" validate:"required"`
+
+	// MetricsKey to authenticate to the jenkins metrics plugin
+	MetricsKey string `yaml:"metricsKey" validate:"required"`
+}
+
+type Monitor struct {
+	Output types.FilteringOutput
+	cancel context.CancelFunc
+	ctx    context.Context
+	// lastBuildSent map (key:job_name and value:build_completion_time) used to keep track of processed builds
+	lastBuildSent map[string]int64
+}
+
+var logger = utils.NewThrottledLogger(log.WithFields(log.Fields{"monitorType": monitorType}), 20*time.Second)
+
+func init() {
+	monitors.Register(&monitorMetadata, func() interface{} { return &Monitor{} }, &Config{})
+}
+
+// Configure monitor
+func (m *Monitor) Configure(conf *Config) error {
+	httpClient, err := conf.HTTPConfig.Build()
+	if err != nil {
+		return err
+	}
+	jkClient, err := jc.NewJenkinsClient(conf.Host, conf.MetricsKey, conf.Port, conf.Scheme(), httpClient)
+	if err != nil {
+		return err
+	}
+	m.ctx, m.cancel = context.WithCancel(context.Background())
+	m.lastBuildSent = make(map[string]int64)
```

Comment:
This makes me wary.  See if the metrics can be sent in such a way (e.g. a cumulative counter or histogram-like set of metrics) that don't require keeping state like this.

Reply:
we'll discuss in person to why this approach is followed